pallets/interest-accrual/src/lib.rs

```diff
@@ -370,40 +387,52 @@ pub mod pallet {
 		}
 
 		pub fn reference_interest_rate(interest_rate_per_sec: T::InterestRate) -> DispatchResult {
-			Rate::<T>::mutate(interest_rate_per_sec, |rate| {
-				if let Some(rate) = rate {
-					rate.reference_count += 1;
-				} else {
-					let rate_count = RateCount::<T>::get();
-					ensure!(
-						rate_count < T::MaxRateCount::get(),
-						Error::<T>::TooManyRates
-					);
-
-					*rate = Some(RateDetailsOf::<T> {
-						accumulated_rate: One::one(),
-						reference_count: 1,
-					});
-					RateCount::<T>::set(rate_count.checked_add(1).ok_or(Error::<T>::TooManyRates)?);
+			Rates::<T>::try_mutate(|rates| {
+				for rate in rates.iter_mut() {
+					if rate.interest_rate_per_sec == interest_rate_per_sec {
+						rate.reference_count += 1;
+						return Ok(());
+					}
 				}
+				// Fell through the loop, so push in a new item
+				let new_rate = RateDetailsOf::<T> {
+					interest_rate_per_sec,
+					accumulated_rate: One::one(),
+					reference_count: 1,
+				};
+				rates
+					.try_push(new_rate)
+					.map_err(|_| Error::<T>::TooManyRates)?;
 				Ok(())
 			})
 		}
 
 		pub fn unreference_interest_rate(interest_rate_per_sec: T::InterestRate) -> DispatchResult {
-			Rate::<T>::try_mutate(interest_rate_per_sec, |maybe_rate| {
-				if let Some(rate) = maybe_rate {
-					rate.reference_count = rate.reference_count.saturating_sub(1);
-					if rate.reference_count == 0 {
-						*maybe_rate = None;
-					}
-					Ok(())
-				} else {
-					Err(Error::<T>::NoSuchRate.into())
+			Rates::<T>::try_mutate(|rates| {
+				let idx = rates
+					.iter()
+					.enumerate()
+					.find(|(_, rate)| rate.interest_rate_per_sec == interest_rate_per_sec)
+					.ok_or(Error::<T>::NoSuchRate)?
+					.0;
+				rates[idx].reference_count = rates[idx].reference_count.saturating_sub(1);
+				if rates[idx].reference_count == 0 {
+					rates.remove(idx);
 				}
+				Ok(())
 			})
 		}
 
+		pub fn get_rate(
+			interest_rate_per_sec: T::InterestRate,
+		) -> Result<RateDetailsOf<T>, DispatchError> {
+			let rate = Rates::<T>::get()
+				.into_iter()
+				.find(|rate| rate.interest_rate_per_sec == interest_rate_per_sec)
+				.ok_or(Error::<T>::NoSuchRate)?;
+			Ok(rate)
```

Comment:
Could: Reduce by one line 😇

```suggestion
			 Rates::<T>::get()
				.into_iter()
				.find(|rate| rate.interest_rate_per_sec == interest_rate_per_sec)
				.ok_or(Error::<T>::NoSuchRate.into())
```